opt/scopeserver/dataserver/utils/search.py

```diff
@@ -0,0 +1,267 @@
+            self.set_search_space(self.loom.hsap_ss)
+            self.set_cross_species("hsap")
+            self.set_search_term(search_term[5:])
+        elif search_term.startswith("mmus\\"):
+            self.set_search_space(self.loom.mmus_ss)
+            self.set_cross_species("mmus")
+            self.set_search_term(search_term[5:])
+        else:
+            self.set_search_space(self.loom.ss)
+            self.set_cross_species("")
+
+    def find_matches(self) -> List[SearchMatch]:
+        """
+        Search for matches in the search space.
+
+        Find keys in the search space where the search term matches (casefolded)
+        the searchable element. Return the match, the matches category and relationship.
+        """
+
+        if len(self.casefold_term) == 1:
+            matches = [x for x in self.search_space.keys() if self.casefold_term == x[0]]
+        else:
+            matches = [x for x in self.search_space.keys() if self.casefold_term in x[0]]
+
+        match_results = []
+
+        for match in matches:
+            if isinstance(self.search_space[match], list):
+                for element in self.search_space[match]:
+                    match_results.append((match, (match[1], element, match[2])))
+            else:
+                match_results.append((match, (match[1], self.search_space[match], match[2])))
+
+        perfect_match: List[SearchMatch] = []
+        good_match: List[SearchMatch] = []
+        bad_match: List[SearchMatch] = []
+
+        for match_result in match_results:
+            # This contains the full match, it's category and their relationship
+            match, result = match_result
+
+            if match[1] == self.search_term or match[0] == self.casefold_term:
+                perfect_match.append(result)
+                continue
+            if (
+                match[1].startswith(self.search_term)
+                or match[1].endswith(self.search_term)
+                or match[0].startswith(self.casefold_term)
+                or match[0].endswith(self.casefold_term)
+            ):
+                good_match.append(result)
+                continue
+            if self.search_term in match[0]:
+                bad_match.append(result)
+                continue
+
+        def sort_matches(match: SearchMatch) -> str:
+            return match[0]
+
+        final_matches = (
+            sorted(perfect_match, key=sort_matches)
+            + sorted(good_match, key=sort_matches)
+            + sorted(bad_match, key=sort_matches)
+        )
+
+        return final_matches
+
+    def aggregate_matches(self, matches: List[SearchMatch]) -> Dict[Tuple[str, str], List[str]]:
+        """
+        Collapse matches based on returned element.
+
+        match[0]: Match from search space
+        match[1]: Category of match
+        match[2]: Relationship
+        """
+
+        aggregated_matches: Dict[Tuple[str, str], List[str]] = OrderedDict()
+
+        if self.cross_species == "":
```

Comment:
I would like this `if`/`else` block to be abstracted a bit more. Essentially it is grouping so `itertools.groupby()` will be useful.

Reply:
Something like:
```python
def grouping_key(match: SearchMatch) -> Tuple[str, str]:
    return (match[1], match[2])

arguments = {
    "": (match for match in matches),
    "hsap": itertools.chain.from_iterable(
        (((droso_gene[0], *grouping_key(match)) for droso_gene
          in self.hsap_to_dmel_mappings[match[0]])
         for match in matches)
    ),
    "mmus": itertools.chain.from_iterable(
        (((droso_gene[0], *grouping_key(match)) for droso_gene
          in self.mmus_to_dmel_mappings[self.search_space[match]])
         for match in matches)
    )
}

groups = itertools.groupby(sorted(arguments[self.cross_species], key=grouping_key), key=grouping_key)
aggregated_matches = {key: [match[0] for match in group] for key, group in groups}
```
2 things for this to be viable:
1. why are the mappings being indexed differently?
2. I don't like how the data are being extracted from the mappings. Needs more thought.

Reply:
Not relevant now the cross species stuff is removed